enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+  to apply the configuration to, as such they must have the same CPU set size.**
+
+## Proposal
+
+We will need to maintain a global identifier that is set during installation and
+can not be easily removed after the fact. This approach will help remove
+exposing this feature via an API and limit the chances that a misconfiguration
+can cause un-recoverable scenarios for our customers. At install time we will
+also apply an initial machine config for workload partitioning that sets a
+default CPUSet for the whole CPUSet. Effectively this will behave as if workload
+partitioning is not turned on. When a customer wishes to pin the management
+workloads they will be able to do that via the existing Performance Profile.
+Resizing partition size will not cause any issues after installation.
+
+With this approach we eliminate the race condition that can occur if we apply
+the machine config after bootstrap via NTO. Since we create a "default" cri-o
+and kubelet configuration that does not specify the CPUSet customers do not have
+to worry about configuring correct bounds for their machines and risk
+misconfiguration.
+
+Furthermore, as machines join the cluster they will have the feature turned on
+before kubelet and the api-server come up as they query the
+`machine-config-server` for their configurations before joining. This also
+allows us more flexibility and an easier interface for the customer since
+customers only need to interact with the Performance Profile to set their
+`reserved` and `isolated` CPUSet. This makes things less prone to error as not
+only can the CPUSets be different for workers and masters but the machines
+themselves might have vastly different core counts.
+
+In order to implement this enhancement we are proposing working on the 4 components
+defined below.
+
+1. Openshift API - ([Infrastructure
+   Status](https://github.com/openshift/api/blob/81fadf1ca0981f800029bc5e2fe2dc7f47ce698b/config/v1/types_infrastructure.go#L51))
+   - The change in this component is to store a global identifier if we have
+     partitioning turned on.
+2. Openshift Installer
+   - Add an `InstallConfig` flag to the installer to enable this feature from
+     install time only.
+3. Machine Config Operator
+   - Add the ability of MCO to generate the needed machine configs for the
+     worker pools from bootstrap and maintain it.
+4. Admission Controller ([management cpus
+   override](https://github.com/openshift/kubernetes/blob/a9d6306a701d8fa89a548aa7e132603f6cd89275/openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/doc.go))
+   in openshift/kubernetes.
+   - This change will be in support of checking the global identifier in order
+     to modify the pod spec with the correct `requests`.
+5. Node Admission Plugin in openshift/kubernetes
+   - We will add an admission plugin for nodes to prevent nodes from joining a
+     cluster that are not correctly setup for CPU Partitioning.
```

Comment:
Right since this will be tied to MCP, a node that joins an MCP will get it's configuration before kubelet comes up. So as nodes get added after day 0, they'll get the correct configuration at that stage. So it's not the default that will be present on those machines, it'll be what ever the user has already setup the partitioning to be.

Reply:
Are you sure about that? The kubelet must first start the MCO daemon to be able to apply the node changes. And MCO daemon reads the info from the rendered MCP via a standard API server connection. So it seems to me the node must join the cluster first.

Reply:
That's a fair question, that is the behavior I have seen through experimentation of this feature and POCing out the flow. That behavior is also defined in in the [machine config server](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigServer.md) component of MCO documentation. So the incoming node would query for it's configs before kubelet comes online and joins the cluster.

Reply:
Right, the admission controller will stop any node from entering the cluster that isn't configured correctly. An admin will need to create a machine pool with out partitioning, and then create a node manifest prior to the kubelet creating it, and alter it to lie to the api-server that it is configured correctly for the admission plugin to be fooled.